File: com.mobi.platform.config.api/src/main/java/com/mobi/platform/config/api/state/StateManager.java
package com.mobi.platform.config.api.state;

/*-
 * #%L
 * com.mobi.platform.config.api
 * $Id:$
 * $HeadURL:$
 * %%
 * Copyright (C) 2016 iNovex Information Systems, Inc.
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * #L%
 */

import com.mobi.rdf.api.Model;
import com.mobi.rdf.api.Resource;

import java.util.Map;
import java.util.Set;
import javax.annotation.Nullable;

public interface StateManager {
    /**
     * Determines whether a State with the passed ID exists in the repository.
     *
     * @param stateId the ID of the State to look for NOTE: Assumes ID represents an IRI unless String
     *      begins with "_:".
     * @return true if the State object exists; false otherwise
     */
    boolean stateExists(Resource stateId);

    /**
     * Determines whether a State with the passed ID is for the User with the passed username.
     *
     * @param stateId the ID of the State to check NOTE: Assumes ID represents an IRI unless String
     *      begins with "_:".
     * @param username the username of a User in Mobi
     * @return true if the State is for the User; false otherwise
     * @throws IllegalArgumentException Thrown if the State or User could not be found
     */
    boolean stateExistsForUser(Resource stateId, String username);

    /**
     * Stores the passed Model in the repository as a new State for the User with the passed username.
     * Every subject in the Model will be added to the new State object.
     *
     * @param newState a collection of statements to link to the new State
     * @param username the username of the User to associate with the new State
     * @return The IRI of the new State
     * @throws IllegalArgumentException Thrown if the User could not be found
     */
    Resource storeState(Model newState, String username);

    /**
     * Stores the passed Model in the repository as a new ApplicationState for the User with the passed
     * username and the application with the passed ID. Every subject in the Model will be added to the
     * new ApplicationState object.
     *
     * @param newState a collection of statements to link to the new State
     * @param username the username of the User to associate with the new State
     * @param applicationId the ID of the Application to attach the new State to
     * @return The IRI of the new ApplicationState
     * @throws IllegalArgumentException Thrown if the User or Application could not be found
     */
    Resource storeState(Model newState, String username, String applicationId);

    /**
     * Removes State with the passed ID from the repository. Removes all associated statements unless
     * their subjects are related to other States.
     *
     * @param stateId the ID of the State to remove from the repository NOTE: Assumes ID represents an
     *      IRI unless String begins with "_:".
     * @throws IllegalArgumentException Thrown if the State could not be found
     */
    void deleteState(Resource stateId);

    /**
     * Updates the State with the passed ID with the passed new Model of statements.
     *
     * @param stateId the ID of the State to update NOTE: Assumes ID represents an IRI unless String
     *      begins with "_:".
     * @param newState the new Model of statements to associate with the State
     * @throws IllegalArgumentException Thrown if the State could not be found
     */
    void updateState(Resource stateId, Model newState);

    /**
     * Retrieves all State IDs with Models of all associated statements for States that match the passed filters.
     * Can filter by the associated User with the passed username, associated application ID, and associated subjects.
     *
     * @param username the optional username of a User to filter State by
     * @param applicationId the optional ID of the Application to filter State by
     * @param subjects a Set of subject IRIs to filter the State by
     * @return a Map of State IDs to Models of all the associated statements for all State matching the filter criteria
     * @throws IllegalArgumentException Thrown if the User or Application could not be found
     */
    Map<Resource, Model> getStates(@Nullable String username, @Nullable String applicationId,
                                   Set<Resource> subjects);

    /**
     * Retrieves a State Model by the passed State ID.
     *
     * @param stateId the ID of the State to retrieve
     * @return a Model of all the statements associated with the specified State
     * @throws IllegalArgumentException Thrown if the State could not be found
     */
    Model getState(Resource stateId);
}

File: com.mobi.web/src/main/resources/public/js/services/stateManager.js
/*-
 * #%L
 * com.mobi.web
 * $Id:$
 * $HeadURL:$
 * %%
 * Copyright (C) 2016 iNovex Information Systems, Inc.
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * #L%
 */
(function() {
    'use strict';

    angular
        /**
         * @ngdoc overview
         * @name stateManager
         *
         * @description
         * The `stateManager` module only provides the `stateManagerService` service which provides access to the Mobi
         * state REST endpoints and all of the current user's state.
         */
        .module('stateManager', [])
        /**
         * @ngdoc service
         * @name stateManager.service:stateManagerService
         * @requires util.service:utilService
         *
         * @description
         * `stateManagerService` is a service that provides access to the Mobi state REST endpoints and the `states`
         * variable which holds all the state for the currently logged in user.
         */
        .service('stateManagerService', stateManagerService);

        stateManagerService.$inject = ['$http', '$q', '$httpParamSerializer', 'utilService', 'REST_PREFIX'];

        function stateManagerService($http, $q, $httpParamSerializer, utilService, REST_PREFIX) {
            var self = this;
            var prefix = REST_PREFIX + 'states';
            var util = utilService;

            /**
             * @ngdoc property
             * @name states
             * @propertyOf stateManager.service:stateManagerService
             * @type {Object[]}
             *
             * @description
             * `states` holds the list of all states for the current user. Each object looks like this:
             * ```
             * {
             *     id: '',
             *     model: [...]
             * }
             * ```
             */
            self.states = [];

            /**
             * @ngdoc method
             * @name initialize
             * @methodOf stateManager.service:stateManagerService
             *
             * @description
             * Initializes the `states` variable using the `getStates` method. If the states cannot be retrieved,
             * creates an error toast.
             *
             * @returns {Promise} A promise that resolves whether the initialization was successful or not
             */
            self.initialize = function() {
                return self.getStates()
                    .then(states => self.states = states, () => util.createErrorToast('Problem getting states'));
            }
            /**
             * @ngdoc method
             * @name getStates
             * @methodOf stateManager.service:stateManagerService
             *
             * @description
             * Calls the GET /mobirest/states endpoint with the provided parameters and returns the array of state
             * objects.
             *
             * @param {Object} stateConfig The query parameters to add to the request
             * @returns {Promise} A promise that resolves to the array of state objects or rejects with an error message
             */
            self.getStates = function(stateConfig) {
                var params = $httpParamSerializer(stateConfig);
                return $http.get(prefix + '?' + params)
                    .then(response => _.get(response, 'data', []), util.rejectError);
            }
            /**
             * @ngdoc method
             * @name createState
             * @methodOf stateManager.service:stateManagerService
             *
             * @description
             * Calls the POST /mobirest/states endpoint with the provided state JSON-LD and a string identifying the
             * application that state will be for. If the state was created successfully, it is added to the `states`
             * array. Returns a Promise.
             *
             * @param {Object[]} stateJson A JSON-LD array of the state data.
             * @param {string} application A string identifying the application the state will belong to
             * @returns {Promise} A promise that resolves if the creation was successful or rejects with an error message
             */
            self.createState = function(stateJson, application) {
                var config = {
                    transformResponse: undefined,
                    headers: {
                        'Content-Type': 'application/json'
                    }
                };
                if (application) {
                    config.params = {application};
                }
                return $http.post(prefix, angular.toJson(stateJson), config)
                    .then(response => self.states.push({id: response.data, model: stateJson}), util.rejectError);
            }
            /**
             * @ngdoc method
             * @name getState
             * @methodOf stateManager.service:stateManagerService
             *
             * @description
             * Calls the GET /mobirest/states/{stateId} endpoint with the provided state IRI string to retrieve a
             * state object. Returns a Promise with the state object.
             *
             * @param {string} stateId A string identifying the state to retrieve
             * @returns {Promise} A promise that resolves with the identified state object or rejects with an error
             *      message
             */
            self.getState = function(stateId) {
                return $http.get(prefix + '/' + encodeURIComponent(stateId))
                    .then(response => _.get(response, 'data', {}), util.rejectError);
            }
            /**
             * @ngdoc method
             * @name updateState
             * @methodOf stateManager.service:stateManagerService
             *
             * @description
             * Calls the PUT /mobirest/states/{stateId} endpoint and updates the identified state with the provided IRI
             * string with the provided JSON-LD array of new data. If the update was successful, updates the `states`
             * array with the new model data. Returns a Promise indicating the success.
             *
             * @param {string} stateId A string identifying the state to retrieve
             * @param {Object[]} stateJson A JSON-LD array of the new state data
             * @returns {Promise} A promise that resolves if the update was successful or rejects with an error message
             */
            self.updateState = function(stateId, stateJson) {
                return $http.put(prefix + '/' + encodeURIComponent(stateId), angular.toJson(stateJson))
                    .then(() => _.forEach(self.states, state => {
                        if (_.get(state, 'id', '') === stateId) {
                            _.set(state, 'model', stateJson);
                            return false;
                        }
                    }), util.rejectError);
            }
            /**
             * @ngdoc method
             * @name deleteState
             * @methodOf stateManager.service:stateManagerService
             *
             * @description
             * Calls the DELETE /mobirest/states/{stateId} endpoint to remove the identified state object with the
             * provided IRI from the application. If the deletion was successful, updates the `states` array. Returns a
             * Promise indicating the success.
             *
             * @param {string} stateId A string identifying the state to delete
             * @returns {Promise} A promise that resolves if the deletion was successful or rejects with an error message
             */
            self.deleteState = function(stateId) {
                return $http.delete(prefix + '/' + encodeURIComponent(stateId))
                    .then(() => _.remove(self.states, {id: stateId}), util.rejectError);
            }
        }
})();
